✨ Send account events to ozone

foysalit opened the pull request:
Specific features:

- automod engine feature for persisting record events (create/update/delete) to Ozone (if a rule requests this)
- rules to track ozone event on record subjects, and then persist future record updates to Ozone
- optional logic to persist account and identity events to ozone

whyrusleeping committed:
WIP: creating bgs splitter daemon

whyrusleeping committed:
splitter is now functional

whyrusleeping committed:
fix edge cases around playback-to-live transition

bnewbold commented:
Yup! This is roughly the shape I was thinking. Some notes:

- should have a conditional that if there isn't an authenticated ozone client configured on the engine, just skip
- an explicit config flag for this behavior. for example, in prod we run two parallel instances of automod, one for blobs and one for all other labels; we would only want to do these lifecycle event pushes from one of those instances
- we could add more context/metadata to the args of `ProcessIdentityEvent`, it is very minimal right now

bnewbold commented:
More quick notes:

* to do lexicon generation in indigo, have the `atproto` repo checked out in parallel. eg, I have `~/code/indigo/` and `~/code/atproto/`, so that the lexicons are at `~/code/atproto/lexicons/`. you can have a branch or patches to the lexicon schema JSON files, doesn't need to be `main` branch. then, in indigo, run `make lexgen`, and then run `make fmt lint test` to check if anything broke. for example, new query params mean existing calling code needs to be updated. in some cases `make cborgen` is also necessary.
* to "de-dupe" event creation, i'd probably do a `queryEvents` on the same subject with the same event type, and check that the most recent event wasn't created in the past 48 hours or so and having the same type? might need some more thought/tuning
* just to call it out, there is a tension here between a model of "here is the account/record state at this point in time" vs "here is what changed (diff)". for a human mod reviewer, "what changed" is probably more meaningful, but the technical architecture isn't built that way. we can try to hack around that, but it is always going to be hard to figure out "what was the state before", which is what is required to figure out "what changed". having a log of "current state" allows a human to look and visually see what changed; for example this is what we do for DID PLC history in the account profile view right now

foysalit commented:
Thanks, these notes help a lot!

> check that the most recent event wasn't created in the past 48 hours
these are quite tricky since profiles may be updated within minutes because people are changing their bio or smth or accounts may be deactivated/reactivated because someone wanted to check how it works etc. I think just setting a shorter window for the last event check, like within the last 30s or smth would cover most legit cases?

>  for a human mod reviewer, "what changed" is probably more meaningful, but the technical architecture isn't built that way

I think we would cover this via record snapshots in ozone. automod would just emit the changes and ozone would keep snapshots so that we have full history for record updates. for account events, these don't matter much, do they? I guess for handle change it does but then we kinda have that history in plc, right?

bnewbold commented:
for dedupe: yup! you are right, we definitely want to capture all those edits. maybe we can de-dupe from the CID or timestamp in the most recent event, instead of just the event type?

for "what changed": PLC is a good example. if only looking at PLC changes, the existing table works well: you can scan the columns and see what changed. in the ozone event history, it is probably going to be a mix of reports, comments, actions/labeling, and then a identity/PLC event in the middle. in that context is isn't easy to see what that one identity event represented though, would need to dig deeper. it is definitely an improvement to have it show up at all, and we can probably figure this out later in either the ozone backend or client.

foysalit commented:
Summarizing recent changes

1. we are now keeping count of `mod-event` and only sending events for following cases IF there's at least 1 mod-event on the record/account
a. post delete event
b. any profile event (update and delete since on create we won't have any mod-events)
2. cleaned up the resolve appeal stuff that got mixed into this from another PR

jazware committed:
Configurable Relay Event Retention (#786)

bnewbold committed:
lexgen: handle 'ref' as procedure input type

bnewbold committed:
lexgen 20241030 (#785)

@brianolson: there was a lexgen issue in here that I manually worked
around. we have a new lexicon which uses a ref directly for the API
endpoint "input" schema, which maybe hasn't be exercised before.

bnewbold committed:
wire up abuse metadata from ozone to automod

haileyok committed:
update

bnewbold committed:
fix segfaults

bnewbold committed:
Add avatar to `AccountMeta` (#779)

smol change. couple of new rules need avatar cids. currently have to
fetch from public api before the rule processes in a separate request
which is meh, since we have the info here already.

bnewbold committed:
support tokenize while keeping common censor chars (#767)

There are some cases where it's useful to tokenize a string while _not_
splitting on some non-letter chars like `#`, `*`, `-`, or `_`.
Unfortunately right now `Tokenize` will split on all of these, making
some matching difficult.

This just adds a second `TokenizeTextSkippingCensorChars` for those
particular use cases. Also adding `TokenizeTextWithRegex`, so that other
cases can be easily covered in the future if they arise.

bnewbold committed:
more automod account metadata (#789)

bnewbold committed:
hepa: fix routine order

bnewbold committed:
hepa: fix routine order (#790)

bnewbold committed:
lexgen: handle 'ref' as procedure input type (#788)

This tries to resolve the situation discovered in
https://github.com/bluesky-social/indigo/pull/785, with the ozone
"upsertSet" procedure.

That lexicon specifies the input schema via reference, instead of an
"object" schema in-place.

This code could use real review, I just banged on it until it worked for
the current lexicon.

bnewbold committed:
make goat record creation work w/o schema defs

bnewbold committed:
make goat pref handling schema-agnostic

bnewbold committed:
goat: schema-flexible record creation, and prefs mgmt (#791)


Fixes: https://github.com/bluesky-social/indigo/issues/784

bnewbold committed:
bigsky: show something helpful when you hit home route

bnewbold committed:
bigsky: show something helpful when you hit home route (#793)

bnewbold committed:
relay: env var for persist dir, and alt env var for data dir

bnewbold committed:
relay: env var for persist dir, and alt env var for data dir (#795)

This is mostly driven by feedback from folks running indie relays.

bnewbold committed:
hepa: wire up ozone consumption correctly

bnewbold committed:
hepa: wire up ozone consumption correctly (#796)

brianolson committed:
simple http client for web did

whyrusleeping committed:
make compaction worker count configurable

whyrusleeping committed:
make compaction worker count configurable (#803)

Want to be able to tweak this manually so we can turn it higher during
off-peak if necessary

whyrusleeping committed:
update to more optimized cborgen

whyrusleeping committed:
update to more optimized cborgen (#804)

This should help memory pressure on all firehose consumers, but
especially on the relay

whyrusleeping committed:
update version stuff

whyrusleeping committed:
copy over some updates from the relay

whyrusleeping committed:
dont duplicate sequence for event func

brianolson committed:
wip, start pebble persister

brianolson committed:
pebble persister

brianolson committed:
fix events tests
add pebble test

brianolson committed:
GC thread with retention and period

brianolson committed:
connect rainbow to pebble persister

brianolson committed:
Merge remote-tracking branch 'origin/feat/splitter-rebase' into feat/splitter-rebase

brianolson committed:
delete dead code

brianolson committed:
PR feedback, use pebble DeleteRange

bnewbold committed:
identity: default dir with 100 max idle conns, and 1sec idle

bnewbold committed:
identity: drop default HTTP timeout from 15s to 10s

bnewbold committed:
identity: drop default DNS timeout from 5s to 3s

brianolson committed:
add gauge spl_active_clients

brianolson committed:
--persist-hours

brianolson committed:
fix err log

brianolson committed:
fix broadcast

bnewbold committed:
identity package default tweaks (#811)

- limits on HTTP idle connections. mostly impacts did:web and HTTPS
handle resolution, which can otherwise result in huge numbers of sockets
hanging around (one per custom domain) when operating at scale. values
chosen should not have too bad an impact on PLC usage, though we might
want to split out separate PLC and non-PLC HTTP clients at some point
- drop some timeouts

brianolson committed:
last seq from pebble

brianolson committed:
log gc sizes

whyrusleeping committed:
cors nonsense

whyrusleeping committed:
allow carstore to use multiple directories, round robin style

foysalit committed:
:rightwards_twisted_arrows: Merge with upstream

whyrusleeping committed:
fixup build

foysalit committed:
:sparkles: Fix breaking changes due to lexicon changes

whyrusleeping committed:
more test fixups

brianolson committed:
reorg config

brianolson committed:
Merge remote-tracking branch 'origin/feat/splitter-rebase' into feat/splitter-rebase

foysalit committed:
:rightwards_twisted_arrows: Merge with upstream

foysalit committed:
:broom: Cleanup unnecessary comment

foysalit committed:
:bug: Fix typo in moderation review state check

foysalit committed:
:bug: Use actual value of review state

foysalit committed:
:sparkles: circuit break escalation and ack actions properly

whyrusleeping committed:
sync get record shouldnt unmarshal

foysalit committed:
:rightwards_twisted_arrows: Merge with upstream

foysalit committed:
:bug: Avoid ack if review is closed

whyrusleeping committed:
allow carstore to use multiple directories, round robin style (#812)

whyrusleeping committed:
add a user cache on the bgs

whyrusleeping committed:
fixup refactor

whyrusleeping committed:
fix lint

whyrusleeping committed:
add a user cache on the bgs (#816)

should cut ~3% of all DB usage on the prod relay

brianolson committed:
add compaction to gc

jazware committed:
Add more metrics to the relay

brianolson committed:
dockerize

jazware committed:
whoops

jazware committed:
Rename series

jazware committed:
Track disk vs meta write durations

jazware committed:
Add more metrics to the relay (#817)

brianolson committed:
reorg config, make --cursor-file

brianolson committed:
no-cursor subscribe

brianolson committed:
Merge remote-tracking branch 'origin/main' into feat/splitter-rebase

bnewbold committed:
Bolson/rainbow (#818)

test gha and dockerize

bnewbold committed:
tiny HACKING entry

bnewbold committed:
rainbow README

bnewbold committed:
tweak dockerfile description

bnewbold committed:
import ordering

bnewbold committed:
have daemon refer to itself as rainbow (not splitter)

bnewbold committed:
some args as env vars

bnewbold committed:
defaults for file paths (and env vars)

bnewbold committed:
drop default persist window to 72hr (matching bigsky)

bnewbold committed:
add root path handler for rainbow

bnewbold committed:
Bnewbold/readme rainbow (#822)

going to self-merge this

bnewbold committed:
rainbow service setup tweaks (#821)

- moving service config towards "rainbow" away from "splitter"
- adds env vars for more config options
- defaults and tweaks for some config options
- add a root path HTTP handler, which is helpful for debugging

jazware committed:
Rewind cursor a bit before reconnection for some relay hosts to smooth over reconnection

whyrusleeping committed:
Rewind cursor a bit before reconnection for some relay hosts to smooth over reconnection (#825)

We see some jerk when the Relay restarts caused by high concurrency of
event processing for some high traffic hosts, where some in-progress
event handlers die while the cursor is advanced ahead of them and the
next event for that repo triggers a catch-up/resync of the repo. This
change rewinds the cursor a bit on reconnect to smooth out that process.

brianolson committed:
stream resync to indexer crawler

whyrusleeping committed:
hack: non-archival relay work

whyrusleeping committed:
config flag

brianolson committed:
add metric indexer_catchup_events_failed, add "how" tag to indexer_catchup_events_enqueued

whyrusleeping committed:
add missing file

brianolson committed:
add --next-crawler to forward requestCrawl calls to

haileyok committed:
configurable circuit breakers

haileyok committed:
fix test

haileyok committed:
default values

whyrusleeping committed:
be okay missing revs

haileyok committed:
one more default value

haileyok committed:
automod: configurable circuit breaker thresholds (#830)

brianolson committed:
tool for relay pds resync, and copy pds state from one relay to another

brianolson committed:
fixes

brianolson committed:
use .HasActiveConnection to re-start crawl on target as needed

brianolson committed:
logging tweak

brianolson committed:
better logging

brianolson committed:
Merge remote-tracking branch 'origin/main' into bolson/requestCrawl-forwarder

brianolson committed:
forward requestCrawl to multiple nexts

brianolson committed:
set crawlRequest http client timeout

bnewbold committed:
add --next-crawler to forward requestCrawl calls to (#829)

bnewbold committed:
fix requestCrawl forwarding

bnewbold committed:
fix requestCrawl forwarding (#835)

brianolson committed:
repo_commits_received_counter for every way a commit finishes

brianolson committed:
fix copy-pasteo

brianolson committed:
1am typo

egonelbre committed:
fix staticcheck SA2000

egonelbre committed:
fix staticcheck S1011

egonelbre committed:
fix unnecssary fmt.Sprintf (staticcheck S1039)

Done automatically with `gofmt -w -r "fmt.Sprintf(s) -> s" .`.

brianolson committed:
repo_commits_result_counter{status=ok}

egonelbre committed:
automod/rules: remove unused fmt import

brianolson committed:
add gauge indexer_catchup_repos; fix bigsky NewIndexer usage

brianolson committed:
add memcached for did cache to relay

bnewbold committed:
fix a segfault when passing on requestCrawl fails

bnewbold committed:
fix a segfault when passing on requestCrawl fails (#841)

my bad! had snuck in that recently, `response` is nil if there is an
error.

whyrusleeping committed:
fixing up a few more tests

whyrusleeping committed:
rm deadcode

whyrusleeping committed:
properly pass sequences for repo account events

jazware committed:
properly pass sequences for repo account events (#843)

brianolson committed:
handle response nil, close response.Body

bnewbold committed:
handle response nil, close response.Body (#844)

whyrusleeping committed:
run more tests on both archive/nonarchive

brianolson committed:
pr feedback: .Shudown() instead of Context

jazware committed:
Handle requestCrawl fanout in Rainbow

jazware committed:
requestCrawl should be a post

brianolson committed:
add gauge bgs_connected_inbound

brianolson committed:
fix Shutdown for test

bnewbold committed:
direct DID syntax validation helper

bnewbold committed:
syntax: fast-path for did:plc parsing

Goal is to avoid using full regex parsing for this common case.
profiling was showing that ParseDID regex has been burning a lot of CPU
in prod.

brianolson committed:
fix did memcached metrics

bnewbold committed:
syntax: ParseDID performance (#853)

we are apparently burning a bunch of CPU on this regex in prod, so let's
bypass for the most common case (valid did:plc).

Would be nice to have a benchmark demonstrating perf gains (or not).

I did some manually testing with the CLI helper and inserted `panic()`
statements to confirm that valid DID PLC stays on fast path, and other
DIDs don't.

Feedback welcome if there is a more idiomatic/performant golang way to
do this... this is still doing unicode stuff, I assume, and maybe
casting to bytes would be faster? though probably doesn't impact the
actual hot path much. `strings.ContainsAny` or `strings.ContainsFunc`
could be used here I guess but :shrug:

bnewbold committed:
backfill: basic PDS repo fetching fallback

bnewbold committed:
updates to PDS backfill; and search integration

bnewbold committed:
add PDS backfill log line

brianolson committed:
SetupSlog
GOLOG_ROTATE_KEEP to keep some old logs, rm older

brianolson committed:
SetIpfsWriter, GOLOG_* env to BSKYLOG_*

foysalit committed:
:rightwards_twisted_arrows: Merge with upstream

whyrusleeping committed:
return written blocks in closing deltasession

bnewbold committed:
backfill: add failback to fetch CAR file from PDS (if relay errors) (#855)

The main motivation here is that non-archival relay is returning a 4xx
error for fetching repos, and we want to try fetching those from the
actual PDS when that happens. This code adds a new code branch when a
relay CAR fetch fails to do an identity lookup to find the account's PDS
instance, and fetches the CAR from there.

For the search code specifically, it re-uses an existing identity
directory, to reduce double-resolution.

This also refactors how fetch URLs are constructed to use just
hostnames.

UPDATE: should probably *not* merge this to `main` until Jaz can review

whyrusleeping committed:
better get-record

whyrusleeping committed:
better get-record (#861)

brianolson committed:
fix RWLock usage

bnewbold committed:
add warn log line for unexpected review state

bnewbold committed:
:bug: Fix typo in moderation review state check (#814)

brianolson committed:
replace github.com/ipfs/go-log with stdlib log/slog

brianolson committed:
also respond to GOLOG_{LOG_LEVEL,LOG_FMT,FILE}

bnewbold committed:
lexicon: nsid rkey type

bnewbold committed:
lexicon: nsid rkey type (#865)

https://github.com/bluesky-social/atproto-website/pull/374

bnewbold committed:
make fmt

bnewbold committed:
make fmt (#867)

cc: @brianolson

bnewbold committed:
syntax: path parsing helper

bnewbold committed:
start agnostic sub-package

bnewbold committed:
add existing 'agnostic' helpers

bnewbold committed:
add agnostic version of repo applyWrites

bnewbold committed:
goat: update to use agnostic helpers

bnewbold committed:
astrolabe: updates to use agnostic API helpers

bnewbold committed:
move Directory interface to separate file

bnewbold committed:
move DID doc code out to separate file

bnewbold committed:
re-organize identity package files (#871)

This is a no-op, just shuffling code around between files.

jazware committed:
Update backfill package to use HTTP/HTTPS for getRepo always

jazware committed:
Update backfill package to use HTTP/HTTPS for getRepo always (#875)

audreyt committed:
Use keyed field in DbCID initialization.

fixes "struct literal uses unkeyed fields" in lint.

lizthegrey committed:
feat(cmd/goat): allow specifying authFactorToken

whyrusleeping committed:
small review updates

whyrusleeping committed:
fixup logging now

whyrusleeping committed:
more logging fixup

whyrusleeping committed:
non-archival relay (#827)

real jank right now, but doesnt keep all the repo data, just does basic
event validation

lizthegrey committed:
respond to feedback

lizthegrey committed:
Merge branch 'main' into patch-1

bnewbold committed:
fix refs to original files

lizthegrey committed:
Update account.go

yeah I'm going to make it consistent

brianolson committed:
slog fix

bnewbold committed:
feat(cmd/goat): allow specifying authFactorToken (#877)

Supports using goat account login with [email 2fa
enabled](https://github.com/bluesky-social/atproto/discussions/2435) so
that users do not need to first disable 2fa to then get an email token
anyways to obtain plc signing tokens.

brianolson committed:
fix merge copy-paste-o

brianolson committed:
drop dead code

ngerakines committed:
feature: Added --pds-host argument to blob export

ngerakines committed:
refactor: Updating variables to follow convention

ngerakines committed:
refactor: Removing env variable because it is a bit ambiguous

devinivy committed:
mst: expose NodeData and TreeEntry types

devinivy committed:
goat: add repo inspect --mst flag to display mst structure

bnewbold committed:
Add PDS host argument to blob export (#884)

This PR introduces the `--pds-host` argument to the `blob export`
command chain.

This was needed in order to export the blobs from an old PDS when by DID
document was pointing to the new one already.

bnewbold committed:
sync get record shouldnt unmarshal (#815)

allows this to work on non bsky.app records

bnewbold committed:
make fmt

bnewbold committed:
goat: resolve just DID

bnewbold committed:
goat: make fmt, and resolution mode for just the DID (#887)

bnewbold committed:
use syntax.ParseRepoPath when appropriate

brianolson committed:
simple proxy of listRepos to rainbow upstream

devinivy committed:
goat: switch repo mst flag to command, iterate on output

bnewbold committed:
bump go version to v1.23

bnewbold committed:
fix slog syntax issues

brianolson committed:
better error message

devinivy committed:
goat: support full cid display and root choice in repo mst command

devinivy committed:
tidy

devinivy committed:
goat: support input from stdin for repo mst command

foysalit committed:
:sparkles: Add new typedefs from ozone api changes

foysalit committed:
:sparkles: Adjust query events calls to reflect api change

bnewbold committed:
better error message (#895)

bnewbold committed:
bump golang version to 1.23; fix slog syntax issues (#893)

This is just periodic maintenance, there isn't anything specific in 1.23
we want/need.

go v1.24 is expected in February 2025; we usually wait a month or two
before upgrading (though we didn't get around to 1.23 until now).

The newer `go vet` turned up some slog syntax issues.

bnewbold committed:
:sparkles: Add new typedefs from ozone api changes (#897)

devinivy committed:
goat: use "-" convention for standard io on repo export and repo mst commands

bnewbold committed:
identity: improve error messages and wrapping

devinivy committed:
tidy

bnewbold committed:
goat: longer timeout for CAR export

devinivy committed:
goat: repo mst command for displaying MST structure (#885)

bnewbold committed:
identity: improve error messages and wrapping (#901)

bnewbold committed:
use syntax.ParseRepoPath when appropriate (#889)

This helper was added recently. Rip out old copy-pasta versions and
start actually using it.

bnewbold committed:
bump timeout even higher (10min)

bnewbold committed:
goat: longer timeout for CAR export (#902)

Closes: https://github.com/bluesky-social/indigo/issues/896

Suspect repo exports to crop up in other places, but can start with this
here.

brianolson committed:
disconnect a pds when we block it

haileyok committed:
add suggestions recid codegen

haileyok committed:
other codegen

bnewbold committed:
hepa: remove deprecated/unused config vars

bnewbold committed:
Merge branch 'main' into bnewbold/automod-ozone-persist

Resolved Conflicts:
  api/ozone/moderationqueryStatuses.go

bnewbold committed:
update queryEvents call

bnewbold committed:
hepa: remove deprecated/unused config vars (#909)

this is cleanup from an earlier refactor to move firehose consumption
over to a separate package.

bnewbold committed:
re-word several things

bnewbold committed:
update ozone persisting rules

bnewbold committed:
Merge branch 'main' into bnewbold/automod-ozone-persist

Resolved Conflicts:
	cmd/hepa/main.go
	cmd/hepa/server.go

bnewbold committed:
fix hepa config

bnewbold committed:
move ozone record op persist to main persist function

bnewbold committed:
tweaks to account event forwarding

bnewbold committed:
fix whitespace

bnewbold committed:
add new ozone event types

bnewbold committed:
improve canonical logging

bnewbold committed:
fix bug with variable redeclaration

bnewbold committed:
indicate that OzoneRecordHistoryPersistRule needs to be in record deletion rules

bnewbold committed:
less verbose rule logging

bnewbold committed:
use actual lex refs strings; improve logging; idiomatic err variable

bnewbold committed:
break out variable for logging

whyrusleeping committed:
fixup build after lex updates

haileyok committed:
add suggestions recid codegen (#907)

bnewbold committed:
Merge branch 'main' into escalate-on-high-churn

bnewbold committed:
less verbose logging

bnewbold committed:
:saprkles: Auto escalate accounts for high follow/like churn (#740)

bnewbold commented:
I was re-working this over in https://github.com/bluesky-social/indigo/pull/910, but i'll just push that work here and close the other one.

Summarizing changes made here:

- synced some recent changes from main
- renamed and shuffled some code around; eg "forward" not "reroute"
- made "record op" persist stuff just another record persist behavior, not something controlled via general event forwarding. so basically controlled by enabling a rule, not configuring hepa
- did keep "persist account history" as a hepa-level flag (not a rule), but also made that flag much more explicit about what it is doing.

And some outstanding issues:

at least in staging, the de-dupe query doesn't seem to work. I get:

```
XRPC ERROR 400: InvalidRequest: Invalid event type
```

when using event types like `tools.ozone.moderation.defs#identityEvent`


another problem i'm seeing with "delete" record ops is that some ozone events require a "strong ref"; but with a deletion there is no CID for the record, so a strongref can't be created.

bnewbold committed:
Merge branch 'main' into ozone-account-events

Manually Resolved Conflicts:
 	automod/engine/persist.go

bnewbold commented:
One path forward with this PR is to split out the account/identity update stuff from the record-op update.

The account/identity stuff should be good to go once the de-dupe "queryEvents" issue is resolved, and functionality is tested to work as expected.

The recordOp stuff in the general case needs a resolution to the strongRef problem for deleted records, same as: https://github.com/bluesky-social/indigo/pull/739
